src/System.Diagnostics.DiagnosticSource/src/HttpCorrelationProtocol.md

```diff
@@ -0,0 +1,83 @@
+
+### Request-Id Format
+`Request-Id` is a string up to 1024 bytes length. It contains only [Base64](https://en.wikipedia.org/wiki/Base64) and "-" (hyphen), "|" (vertical bar), "." (dot), and "_" (underscore) characters.
+
+Vertical bar, dot and underscore are reserved characters that used to mark and delimit hierarchical Request-Id, and must not be present in the nodes. Hyphen may be used in the nodes.
+
+Implementations SHOULD support hierarchical structure for the Request-Id, described in [Hierarchical Request-Id document](HierarchicalRequestId.md).
+See [Flat Request-Id](FlatRequestId.md) for non-hierarchical Request-Id requirements.
+
+## Correlation-Context
+Root service can add state (key value pairs) that will automatically propagate to all other services including intermediary services (that support this protocol). A typical scenarios are sampling or A/B testing (feature flags) so that the root service has a way to pass this kind of information down to all services (including intermediary). All services other than the root SHOULD consider Correlation-Context as read-only.
+
+We anticipate that there will be common well-known Correlation-Context keys. If you wish to use this for you own custom (not well-known) context key, prefix it with "@".
+
+It is important to keep the size of any property because they get serialized in HTTP headers and headers have significant size restrictions. The Correlation-Context MUST NOT be used as generic data passing mechanism (between services or within service).
+
+`Correlation-Context` is optional, which means that it may or may not be provided by upstream service.
+
+If `Correlation-Context` is provided by upstream service, implementation MUST propagate it further to downstream services.
+
+Implementation MUST provide read access to `Correlation-Context` for logging systems and MUST support adding properties to Correlation-Context.
+
+### Correlation-Context Format
+`Correlation-Context` is represented as comma separated list of key value pairs, where each pair is represented in key=value format:
+
+`Correlation-Context: key1=value1, key2=value2`
+
+Neither keys nor values MUST NOT contain "=" (equals) or "," (comma) characters. 
+
+Overall Correlation-Context length MUST NOT exceed 1024 bytes, key and value length should stay well under the combined limit of 1024 bytes.
+
+Note that uniqueness of the key within the Correlation-Context is not guaranteed. Context received from upstream service is read-only and implementation MUST not remove or aggregate duplicated keys. 
+
+# HTTP Guidelines and Limitations
+- [HTTP 1.1 RFC2616](https://tools.ietf.org/html/rfc2616)
+- [HTTP Header encoding RFC5987](https://tools.ietf.org/html/rfc5987)
+- De-facto overall HTTP headers size is limited to several kilobytes (depending on a web server)
+
+# Industry standards
```

Comment:
@lmolkova it's part of a broader proposal that a couple parties including Zipkin and Google are working on for distributed trace propagation and submission. We're planning on reaching out more widely once we have more documentation and code samples available, but I'd be glad to lay out the plan and longer-term goals if you're interested.